File: app/src/main/java/com/hammwerk/wizardpager/core/Branch.java
package com.hammwerk.wizardpager.core;

import java.util.Arrays;
import java.util.List;

public class Branch {
	private final List<Page> pages;
	private final String name;

	public Branch(Page... pages) {
		this(null, pages);
	}

	public Branch(String name, Page... pages) {
		if (pages.length == 0) {
			throw new EmptyBranchException();
		}
		for (int i = 0; i < pages.length - 1; i++) {
			if (pages[i] instanceof BranchPage) {
				throw new BranchPageBeforeLastPageException();
			}
		}
		this.pages = Arrays.asList(pages);
		this.name = name;
	}

	public Page getPage(int position) {
		return pages.get(position);
	}

	public int getNumberOfPages() {
		return pages.size();
	}

	public BranchPage getBranchPage() {
		Page page = pages.get(pages.size() - 1);
		if (page instanceof BranchPage) {
			return (BranchPage) page;
		}
		return null;
	}

	public List<Page> getPages() {
		return pages;
	}

	public class EmptyBranchException extends RuntimeException {
	}

	public class BranchPageBeforeLastPageException extends RuntimeException {
	}
}

File: lib/src/main/kotlin/com/hammwerk/wizardpager/core/Branch.kt
package com.hammwerk.wizardpager.core

import android.os.Parcel
import android.os.Parcelable

class Branch(val name: String? = null) : Parcelable {
	constructor(parcel: Parcel?) : this(parcel?.readString()) {
		pages = parcel?.readParcelableArray(Branch::class.java.classLoader)?.map { it as Page }?.toTypedArray()!!
	}

	var pages: Array<out Page> = arrayOf()
		set(value) {
			if (value.dropLast(1).any { it is BranchPage }) {
				throw PageAfterBranchPageException()
			}
			field = value
		}

	val numberOfPages: Int
		get() = pages.size

	val numberOfValidPages: Int
		get() = pages.takeWhile { it.valid }.count()

	val branchPage: BranchPage?
		get() {
			if (pages.isNotEmpty()) {
				val page = pages.last()
				if (page is BranchPage) {
					return page
				}
			}
			return null
		}

	override fun writeToParcel(dest: Parcel?, flags: Int) {
		dest?.writeParcelableArray(pages, 0)
	}

	override fun describeContents(): Int {
		return 0
	}

	inline fun <reified T : Page> getPage(position: Int): T {
		when {
			position < pages.size -> return pages[position] as T
			else -> throw PageIndexOutOfBoundsException()
		}
	}

	companion object {
		@JvmField val CREATOR = object : Parcelable.Creator<Branch> {
			override fun createFromParcel(source: Parcel?): Branch {
				return Branch(source)
			}

			override fun newArray(size: Int): Array<out Branch?> {
				return arrayOfNulls(size)
			}
		}
	}

	class PageAfterBranchPageException : RuntimeException()
}
